File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k % n;
        reverse(0, n-1-k, nums); // reverse the starting (total -k) positions first as we need to shift right
        reverse(n-k, n-1, nums); // then k positions 
        reverse(0, n-1, nums); // reverse full array - now to place everything at correct position
        // this way we dont use extra space 
    }
    public void reverse(int start, int end, int[] nums){
        while (start <= end)
        {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
}

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        length = len(nums)
        k %= length

        start, count = 0, 0

# the total count of numbers exclusive numbers placed at their correct position will be k× n / k = n

        while count < length: 
            current_idx, prev_val = start, nums[start]
            while True:
                new_idx = (current_idx + k) % length
        # place the old value at new index and keep copy of this index value for next iteration
                nums[new_idx], prev_val = prev_val, nums[new_idx] 
                # print("new state of nums: ", nums)
                current_idx = new_idx
                count += 1

                if current_idx == start: # means placed all values at required position for this value of start 
                    break
            
            start += 1

# [1,2,3,4,5,6] k = 2
# new state of nums:  [1, 2, 1, 4, 5, 6]
# new state of nums:  [1, 2, 1, 4, 3, 6]
# new state of nums:  [5, 2, 1, 4, 3, 6]
# new state of nums:  [5, 2, 1, 2, 3, 6]
# new state of nums:  [5, 2, 1, 2, 3, 4]
# new state of nums:  [5, 6, 1, 2, 3, 4]